src/libraries/System.Memory/tests/Span/EnumerateRunePositions.cs

```diff
@@ -0,0 +1,248 @@
+        [InlineData(new byte[] { 0x47, 0xF4, 0x90, 0x57 }, new int[] { 0x47, 0xFFFD, 0xFFFD, 0x57, })] // [ F4 90 ] would place us beyond U+10FFFF, all such sequences have maximal invalid subsequence length 1
+        [InlineData(new byte[] { 0x48, 0xE2, 0x88, 0xB4, 0x58 }, new int[] { 0x48, 0x2234, 0x58, })] // [ E2 88 B4 ] is U+2234 THEREFORE
+        [InlineData(new byte[] { 0x49, 0xE2, 0x88, 0xC0, 0x59 }, new int[] { 0x49, 0xFFFD, 0xFFFD, 0x59, })] // [ E2 88 ] followed by non-continuation byte, maximal invalid subsequence length 2
+        [InlineData(new byte[] { 0x50, 0xF0, 0x9F, 0x98, 0x60 }, new int[] { 0x50, 0xFFFD, 0x60, })] // [ F0 9F 98 ] is valid 3-byte start of 4-byte sequence
+        [InlineData(new byte[] { 0x51, 0xF0, 0x9F, 0x98, 0x20, 0x61 }, new int[] { 0x51, 0xFFFD, 0x20, 0x61, })] // [ F0 9F 98 ] followed by non-continuation byte, maximal invalid subsequence length 3
+        [InlineData(new byte[] { 0x52, 0xF0, 0x9F, 0x98, 0xB2, 0x62 }, new int[] { 0x52, 0x1F632, 0x62, })] // [ F0 9F 98 B2 ] is U+1F632 ASTONISHED FACE
+        public static void EnumerateRunePositions_Battery8(byte[] bytes, int[] expected)
+        {
+            List<int> enumeratedValues = new List<int>();
+
+            RunePosition.Utf8Enumerator enumerator = RunePosition.EnumerateUtf8(bytes).GetEnumerator();
+            Assert.Equal(default, enumerator.Current);
+
+            int expectedIndex = 0;
+            while (enumerator.MoveNext())
+            {
+                bool wasReplaced = Rune.DecodeFromUtf8(bytes.AsSpan(expectedIndex), out Rune expectedRune, out int charsConsumed) != OperationStatus.Done;
+                RunePosition runePosition = enumerator.Current;
+
+                RunePosition_TestProps(expectedRune, expectedIndex, charsConsumed, wasReplaced, runePosition);
+
+                expectedIndex += charsConsumed;
+                enumeratedValues.Add(runePosition.Rune.Value);
+            }
+            Assert.Equal(default, enumerator.Current);
+
+            Assert.Equal(expected, enumeratedValues.ToArray());
+        }
+
+        [Fact]
+        public static void EnumerateRunePositions_DoesNotReadPastEndOfSpan()
+        {
+            // As an optimization, reading scalars from a string *may* read past the end of the string
+            // to the terminating null. This optimization is invalid for arbitrary spans, so this test
+            // ensures that we're not performing this optimization here.
+
+            {
+                ReadOnlySpan<char> span = "xy\U0002B754z".AsSpan(1, 2); // well-formed string, but span splits surrogate pair
+
+                List<int> enumeratedValues = new List<int>();
+                foreach (RunePosition runePosition in RunePosition.EnumerateUtf16(span))
+                {
+                    enumeratedValues.Add(runePosition.Rune.Value);
+                }
+                Assert.Equal(new int[] { 'y', '\uFFFD' }, enumeratedValues.ToArray());
+            }
+
+            {
```

Comment:
The "optimization while reading scalars might rely on the terminating null" concern does not apply to UTF-8. You can delete this entire block.